File: SparkScala/src/Preprocessing.scala
import de.halcony.argparse.Parser
import org.apache.log4j.{Level, Logger}
import org.apache.spark.sql.execution.streaming.FileStreamSource.Timestamp
import org.apache.spark.sql.functions.{from_unixtime, to_date, to_timestamp, udf}
import org.apache.spark.sql.{Dataset, SparkSession}

object Preprocessing {

  private val logger: Logger = Logger.getLogger("app")

  def main(args: Array[String]): Unit = {

    setUpLogging()

    val arguments = parseCLIArguments(args)

    val spark = getSpark(arguments)

    process(arguments("viewsFilepath"), arguments("repartitionedViewsFilepath"), arguments("purchasesFilepath"), arguments("repartitionedPurchasesFilepath"), spark)

    spark.close()
  }

  def setUpLogging(): Unit = {
    Logger.getLogger("org").setLevel(Level.ERROR)
  }

  /**
   * Parses cli arguments
   *
   * @param args array with cli arguments
   * @return Map[String, String] of parsed cli arguments
   */
  def parseCLIArguments(args: Array[String]): Map[String, String] = {
    val parser = Parser("Data preprocessor", "App that preprears data for processing")
      .addPositional("mode", "Mode in which app will run (dev or prod)")
      .addPositional("master", "Spark master")
      .addPositional("viewsFilepath", "Path to a file with views")
      .addPositional("purchasesFilepath", "Path to a file with purchases")
      .addPositional("repartitionedViewsFilepath", "Path to a place where to write views")
      .addPositional("repartitionedPurchasesFilepath", "Path to a place where to write purchases")
      .addOptional("serviceAccountKeyFilepath", "SAKF", description = "Path to a file wiht service account credentials")
      .addOptional("serviceAccountEmail", "SAE", description = "Email address of a service account")
    val arguments: Map[String, AnyVal] = parser.parse(args).toMap
    arguments.map(x => (x._1, String.valueOf(x._2)))
  }


  /**
   * Builds and sets up spark session for desired deployment mode
   *
   * @param args parsed cli arguments
   * @return spark session set up for desired deployment mode
   */
  def getSpark(args: Map[String, String]): SparkSession = {
    args("mode") match {
      case "dev" => getSparkForDevelopment(args("master"))
      case "prod" => getSparkForProduction(args("master"), args("serviceAccountKeyFilepath"), args("serviceAccountEmail"))
      case _ => throw new IllegalArgumentException("mode can be dev or prod")
    }
  }

  /**
   * Builds and sets up spark session for development mode
   *
   * @param master spark master
   * @return spark session set up for development mode
   */
  def getSparkForDevelopment(master: String): SparkSession = {
    SparkSession.builder().master(master).appName("Preprocessing app").getOrCreate()
  }

  /**
   * Builds and sets up spark session for the production mode
   *
   * @param master                    spark master
   * @param serviceAccountKeyFilepath filepath to the key of GCP service account
   * @param serviceAccountEmail       email of the GCP service account
   * @return spark session set up for the production mode
   */
  def getSparkForProduction(master: String, serviceAccountKeyFilepath: String, serviceAccountEmail: String): SparkSession = {
    val spark: SparkSession = SparkSession.builder().appName("Preprocessing app").master(master).getOrCreate()
    setUpServiceAccountCredentials(serviceAccountKeyFilepath, serviceAccountEmail, spark)
    spark
  }
  //TODO: add articles to words :)
  // TODO: ask about comentaries that repeate name of the function

  /**
   * Configures service account in a spark session
   *
   * @param serviceAccountKeyFilepath filepath to key of the GCP service account
   * @param serviceAccountEmail       email of the GCP srvice account
   * @param spark                     spark session to configure
   */
  def setUpServiceAccountCredentials(serviceAccountKeyFilepath: String, serviceAccountEmail: String, spark: SparkSession): Unit = {
    spark.conf.set("google.cloud.auth.service.account.enable", value = true)
    spark.conf.set("google.cloud.auth.service.account.email", serviceAccountEmail)
    spark.conf.set("google.cloud.auth.service.account.keyfile", serviceAccountKeyFilepath)

  }

  /**
   * @param viewsFilepath                  path to a csv file that contains views
   * @param repartitionedViewsFilepath     path where to write repartitioned views
   * @param purchasesFilepath              path to a json file that contains purchases
   * @param repartitionedPurchasesFilepath path where to write repartitioned purchases
   * @param spark                          spark session
   */
  def process(viewsFilepath: String, repartitionedViewsFilepath: String, purchasesFilepath: String, repartitionedPurchasesFilepath: String, spark: SparkSession): Unit = {
    processViews(viewsFilepath, repartitionedViewsFilepath, spark)
    processPurchases(purchasesFilepath, repartitionedPurchasesFilepath, spark)
  }

  def processViews(viewsFilepath: String, repartitionedViewsFilepath: String, spark: SparkSession): Unit = {
    logger.info("Started loading views")
    val views = loadViews(viewsFilepath, spark)
    logger.info("Finished loading views")

    logger.info("Started views repartitioned")
    val repartitionedViews = repartitionView(views, spark)
    logger.info("Finished views repartitioned")

    logger.info("Started writing views")
    repartitionedViews.write.parquet(repartitionedViewsFilepath)
    logger.info("Finished writing views")

  }

  def loadViews(filepath: String, spark: SparkSession): Dataset[View] = {
    import spark.implicits._
    spark.read
      .option("header", "true")
      .csv(filepath)
      .select($"user_id", $"device", $"ip", $"item_id", to_timestamp($"ts").as("ts"))
      .as[View]
  }

  def repartitionView(views: Dataset[View], spark: SparkSession): Dataset[View] = {
    import spark.implicits._
    views.repartitionByRange(to_date($"ts"))
  }

  def processPurchases(purchasesFilepath: String, repartitionedPurchasesFilepath: String, spark: SparkSession): Unit = {
    logger.info("Started loading purchases")
    val purchases = loadPurchases(purchasesFilepath, spark)
    logger.info("Finished loading purchases")

    logger.info("Started purchases repartitioned")
    val repartitionedPurchases = repartitionPurchases(purchases, spark)
    logger.info("Finished purchases repartitioned")

    logger.info("Started writing purchases")
    repartitionedPurchases.write.parquet(repartitionedPurchasesFilepath)
    logger.info("Finished writing purchases")

  }

  def loadPurchases(filepath: String, spark: SparkSession): Dataset[Purchase] = {
    import spark.implicits._
    val parse = udf((values: Array[Array[String]]) => for (value <- values) yield (value(0), value(1).toInt))
    spark.read
      .json(filepath)
      .select($"order_id", $"user_id", $"ip",
        parse($"items").as("items"), to_timestamp(from_unixtime($"ts" / 1000, "yyyy-MM-dd HH:mm:ss.SSSS")).as("ts"))
      .as[Purchase]
  }

  def repartitionPurchases(purchases: Dataset[Purchase], spark: SparkSession): Dataset[Purchase] = {
    import spark.implicits._
    purchases.repartitionByRange(to_date($"ts"))
  }

  case class View(item_id: String, user_id: String, ip: String, ts: Timestamp)

  case class Purchase(order_id: String, user_id: String, ip: String, items: Array[(String, Int)], ts: Timestamp)
}

File: PythonSpark/src/Preprocessing.py
import argparse
import logging

import pyspark.sql
from pyspark.sql import SparkSession
from pyspark.sql.functions import column, to_timestamp, to_date, udf, from_unixtime


def parse_cli_arguments() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description='Preprocessing')
    parser.add_argument('mode', help='Mode in which app will run (dev or prod)', type=str)
    parser.add_argument('master', help='Spark master', type=str)
    parser.add_argument('views_filepath', help='Path to a file with views', type=str)
    parser.add_argument('purchases_filepath', help='Path to a file with purchases', type=str)
    parser.add_argument('repartitioned_views_filepath', help='Path to a place where to write views', type=str)
    parser.add_argument('repartitioned_purchases_filepath', help='Path to a place where to write purchases', type=str)
    parser.add_argument('--SAKF', dest='service_account_key_filepath',
                        help='Path to a file with service account credentials', type=str)
    parser.add_argument('--SAE', dest='service_account_email', help='Email address of a service account', type=str)

    return parser.parse_args()


def get_spark_session_for_development(master: str) -> SparkSession:
    return SparkSession.builder.master(master).appName("Preprocessing app").getOrCreate()


def set_up_service_account(service_account_email: str, service_account_key_filepath: str, spark: SparkSession) -> None:
    spark.conf.set("google.cloud.auth.service.account.enable", "true")
    spark.conf.set("google.cloud.auth.service.account.email", service_account_email)
    spark.conf.set("google.cloud.auth.service.account.keyfile", service_account_key_filepath)


def get_spark_session_for_production(master: str, service_account_key_filepath: str,
                                     service_account_email: str) -> SparkSession:
    spark = SparkSession.builder.master(master).appName("Preprocessing app").getOrCreate()
    set_up_service_account(service_account_email, service_account_key_filepath, spark)
    return spark


def get_spark_session(args: argparse.Namespace) -> SparkSession:
    if args.mode == 'dev':
        return get_spark_session_for_development(args.master)
    elif args.mode == 'prod':
        return get_spark_session_for_production(args.master, args.service_account_key_filepath,
                                                args.service_account_email)
    else:
        raise ValueError("Deployment mode can be dev or prod")


def load_views(views_filepath: str, spark: SparkSession) -> pyspark.sql.DataFrame:
    raw_views = spark.read.option("header", "true").csv(views_filepath)
    parsed_views = raw_views.select(column("user_id"), column("device"), column("ip"),
                                    to_timestamp(column("ts")).alias("ts"))
    return parsed_views


def repartition_views(views: pyspark.sql.DataFrame) -> pyspark.sql.DataFrame:
    return views.repartition(to_date(column("ts")))


def process_views(views_filepath: str, repartitioned_views_filepath: str, spark: SparkSession) -> None:
    logging.info("Started loading views")
    views = load_views(views_filepath, spark)
    logging.info("Finished loading views")

    logging.info("Started views repartitioning")
    repartitioned_views = repartition_views(views)
    logging.info("Finished views repartitioning")

    logging.info("Started writing views")
    repartitioned_views.write.parquet(repartitioned_views_filepath)
    logging.info("Finished writing views")


def load_purchases(purchases_filepath: str, spark: SparkSession) -> pyspark.sql.DataFrame:
    raw_purchases = spark.read.json(purchases_filepath)
    parse_items = udf(lambda raw_list: [[element[0], int(element[1])] for element in raw_list])
    return raw_purchases.select(column("user_id"),
                                column("ip"),
                                parse_items(column("items")).alias("items"),
                                to_timestamp(from_unixtime(column("ts") / 1000, "yyyy-MM-dd HH:mm:ss.SSSS")).alias("ts"))


def repartition_purchases(purchases: pyspark.sql.DataFrame) -> pyspark.sql.DataFrame:
    return purchases.repartition(to_date(column("ts")))


def process_purchases(purchases_filepath: str, repartitioned_purchases_filepath: str, spark: SparkSession) -> None:
    logging.info("Started loading purchases")
    purchases = load_purchases(purchases_filepath, spark)
    logging.info("Finished loading purchases")

    logging.info("Started purchase repartitioning")
    repartitioned_purchases = repartition_purchases(purchases)
    logging.info("Finished purchase repartitioning")

    logging.info("Started writing purchases")
    repartitioned_purchases.write.parquet(repartitioned_purchases_filepath)
    logging.info("Finished writing purchases")


def process(views_filepath: str, repartitioned_views_filepath: str, purchases_filepath: str,
            repartitioned_purchases_filepath: str, spark: SparkSession) -> None:
    process_views(views_filepath, repartitioned_views_filepath, spark)
    process_purchases(purchases_filepath, repartitioned_purchases_filepath, spark)


def set_up_logging() -> None:
    logging.basicConfig(format='%(asctime)s - %(levelname)s [%(name)s] [%(funcName)s():%(lineno)s] - %(message)s',
                        level=logging.INFO)


def __main__():
    args = parse_cli_arguments()

    spark = get_spark_session(args)

    process(args.views_filepath, args.repartitioned_views_filepath,
            args.purchases_filepath, args.repartitioned_purchases_filepath,
            spark)

    spark.stop()


if __name__ == '__main__':
    __main__()
